zebra-state/src/service/finalized_state/disk_format/transparent.rs

```diff
@@ -221,7 +221,15 @@ impl<C: Constraint + Copy + std::fmt::Debug> std::ops::Add for AddressBalanceLoc
         Ok(AddressBalanceLocationInner {
             balance: (self.balance + rhs.balance)?,
             received: self.received.saturating_add(rhs.received),
-            location: self.location.min(rhs.location),
+            // Keep in mind that `AddressBalanceLocationChange` reuses this type
+            // (AddressBalanceLocationInner) and this addition method. The
+            // `block_info_and_address_received` database upgrade uses the
+            // empty/zero location as a dummy value for
+            // `AddressBalanceLocationChange`. Therefore, when adding, we should
+            // ignore these dummy values (which are 0). Using `max` achieves
+            // this. It is also possible that two zero-location balance changes
+            // are added, and `max` will correctly keep the zero value.
+            location: self.location.max(rhs.location),
```

Comment:
We _could_ make this work, but it looks like a foot-gun to me.

It would be better to update `AddressBalanceLocationChange::empty()` to use the max `TransactionLocation` in the add block info and address received balances upgrade.


```rust

impl AddressBalanceLocationChange {
    /// Creates a new [`AddressBalanceLocationChange`] with all zero values and a dummy (all one bits) location.
    fn empty() -> Self {
        Self::new(AddressLocation::from_usize(
            crate::MAX_ON_DISK_HEIGHT,
            usize::MAX,
            usize::MAX,
        ))
    }
}

```

```suggestion
            location: self.location.min(rhs.location),
```

Reply:
I agree. (For anyone reading, the reason it's a foot-gun is because we use the mininum UTXO location for the "canonical" location of an address. We might rely on `min` to preserve that minimum location when adding a AddressBalanceLocationChange to one that might have a (larger) non-canonical location.

Changed in eee60725b